docs/housing-counselor-data.md

```diff
@@ -0,0 +1,163 @@
+
+
+## Fetch agency listings
+
+(_in `fetcher.py`_)
+
+Request every housing counselor in HUD's database with this request:
+
+```
+https://data.hud.gov/Housing_Counselor/searchByLocation?Lat=38.8951&Long=-77.0367&Distance=5000
+```
+
+It returns thousands of results like this:
+
+```json
+    {
+        "services": "DFC,FBC,PPC,RHC",
+        "languages": "ENG,SPA",
+        "agc_STATUS": "A",
+        "agc_SRC_CD": "HUD",
+        "counslg_METHOD": "Face to Face Counseling,Group Counseling,Internet Counseling,Phone Counseling",
+        "agcid": "80790",
+        "adr1": "24331 W. Eight Mile Road",
+        "adr2": " ",
+        "city": "DETROIT",
+        "email": "tmorris@abayomicdc.org",
+        "fax": "313-541-3820",
+        "nme": "ABAYOMI COMMUNITY DEVELOPMENT CORPORATION",
+        "phone1": "313-541-9828-222",
+        "statecd": "MI",
+        "weburl": "www.abayomicdc.org",
+        "zipcd": "48219-1028",
+        "agc_ADDR_LATITUDE": "42.442658",
+        "agc_ADDR_LONGITUDE": "-83.28329",
+        "parentid": "81228",
+        "county_NME": "",
+        "phone2": " ",
+        "mailingadr1": "24331 W. Eight Mile Road",
+        "mailingadr2": " ",
+        "mailingcity": "DETROIT",
+        "mailingzipcd": "48219-1028",
+        "mailingstatecd": "MI",
+        "state_NME": "Michigan",
+        "state_FIPS_CODE": null,
+        "faithbased": "Y",
+        "colonias_IND": "Y",
+        "migrantwkrs_IND": "N"
+    },
+```
+
+## Clean the results
+
+(_in `fetcher.py`_)
+
+We replace the 'languages' value with the full language names. We replace the 'services' value with fully spelled out service descriptions. Both of these mappings come from the HUD API.
+
+(_in `cleaner.py`_)
+
+Clean the data: Convert latitude and longitude values to Floats. Convert the the city and organization name to title case. Ensure email appears to be an email (otherwise leave it blank). Add `http://` to the URL if it is present.
+
+## Backfill missing latitude and longitude values
+
+(_in `geocoder.py`_)
+
+If any counselor records are missing their latitude or longitude data, we fill in those values with the lat/long location of the agency's ZIP code. This uses a saved data file with mappings from each ZIP code to its latitude and longitude values. We create the file of mappings ahead of time, using the `hud_geocode_zipcodes` management command.
+
+## Create collections of results by ZIP code
+
+(_in `generator.py`_)
+
+Create a local sqlite database and define a `distance_in_miles` function in it. Fill the database with a three-column table: For each counselor in the list, include its latitude and longitude (both in radians) and all of its information from the HUD API as a JSON text string.
+
+For each ZIP code in the US, query the database to find the 10 closest housing counselors to the lat/long of that ZIP. Put the information in a JSON structure like this (but with ten results instead of two):
```

Comment:
```suggestion
 For each ZIP code in the U.S., query the database to find the 10 closest housing counselors to the lat/long of that ZIP. Put the information in a JSON structure like this (but with ten results instead of two):
```